sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala

```diff
@@ -85,10 +85,42 @@ object OrcUtils extends Logging {
   }
 
   private def toCatalystSchema(schema: TypeDescription): StructType = {
+    import TypeDescription.Category
+
+    def toCatalystType(orcType: TypeDescription): DataType = {
+      orcType.getCategory match {
+        case Category.STRUCT => toStructType(orcType)
+        case Category.LIST => toArrayType(orcType)
+        case Category.MAP => toMapType(orcType)
+        case _ => CatalystSqlParser.parseDataType(orcType.toString)
+      }
+    }
+
+    def toStructType(orcType: TypeDescription): StructType = {
+      val fieldNames = orcType.getFieldNames.asScala
+      val fieldTypes = orcType.getChildren.asScala
+      fieldNames.zip(fieldTypes).foldLeft(new StructType) {
+        case (resultType, (fieldName, fieldType)) =>
+          val catalystType = toCatalystType(fieldType)
+          resultType.add(StructField(fieldName, catalystType))
+      }
+    }
+
+    def toArrayType(orcType: TypeDescription): ArrayType = {
+      val elementType = orcType.getChildren.asScala.head
```

Comment:
```suggestion
      val elementType = orcType.getChildren(0)
```

Reply:
then `.get(0)`?

Reply:
Exactly. I have no idea why I didn't notice that...